Sia.Data.Incident.Tests/Filters/EventFiltersTests.cs

```diff
@@ -0,0 +1,296 @@
+            {
+                IncidentId = 1,
+                EventTypes = new long[] { 1 },
+                Occurred = new DateTime(1970, 1, 1),
+                EventFired = new DateTime(1970, 1, 1)
+            };
+            var testInput = new List<Event>()
+            {
+                new Event()
+                {
+                    IncidentId = 1,
+                    EventTypeId = 1,
+                    Occurred = new DateTime(1970, 1, 1),
+                    EventFired = new DateTime(1970, 1, 1),
+                    Data = "firstExpectedEvent"
+                },
+                new Event()
+                {
+                    IncidentId = 1,
+                    EventTypeId = 1,
+                    Occurred = new DateTime(1970, 1, 1),
+                    EventFired = new DateTime(1970, 1, 1),
+                    Data = "secondExpectedEvent"
+                },
+                new Event()
+                {
+                    IncidentId = 2,
+                    EventTypeId = 1,
+                    Occurred = new DateTime(1970, 1, 1),
+                    EventFired = new DateTime(1970, 1, 1)
+                },
+                new Event()
+                {
+                    IncidentId = 1,
+                    EventTypeId = 2,
+                    Occurred = new DateTime(1970, 1, 1),
+                    EventFired = new DateTime(1970, 1, 1)
+                },
+                new Event()
+                {
+                    IncidentId = 1,
+                    EventTypeId = 1,
+                    Occurred = new DateTime(1972, 2, 2),
+                    EventFired = new DateTime(1970, 1, 1)
+                },
+                new Event()
+                {
+                    IncidentId = 1,
+                    EventTypeId = 1,
+                    Occurred = new DateTime(1970, 1, 1),
+                    EventFired = new DateTime(1972, 2, 2)
+                },
+            }.AsQueryable();
+
+
+            var result = serviceUnderTest.Filter(testInput);
+
+
+            Assert.AreEqual(2, result.Count());
+            Assert.IsTrue(result.Select(ev => ev.Data).Contains("firstExpectedEvent"));
+            Assert.IsTrue(result.Select(ev => ev.Data).Contains("secondExpectedEvent"));
+        }
+
+        [TestMethod]
+        public void Filter_WhenPassedQueryable_MatchesByDataKeyWhenEventsHaveEquivalentKeyInData()
+        {
+            var serviceUnderTest = new EventFilters()
+            {
+                DataKey = "HelloWorld"
+            };
+            var expectedEventTypeId = 2;
+            var unexpectedEventTypeId = 1;
+            var testInput = new List<Event>()
+            {
+                new Event()
+                {
+                    IncidentId = 1,
+                    EventTypeId = unexpectedEventTypeId,
+                    Occurred = new DateTime(1970, 1, 1),
+                    EventFired = new DateTime(1970, 1, 1),
+                    Data = "NotMatched"
```

Comment:
Filtering needs to happen in the database, not on the Gateway. Not only would doing the filtering on the Gateway use significantly more network and CPU resources, it would make paginated filtering extremely difficult.

Reply:
Is line 46 (.WithFilter(request.Filter) ) in GetEvents.cs implemented in SQL, or the filter is applied after data retrieved?

Reply:
var serviceUnderTest = new EventFilters()
{
    EventTypes = new long[] { 1, 10 },
    DataKey = "HelloWorld",
    DataValue = "HelloWorld"
};

var filteredQuery = dbContextToQuery.Events.WithFilter(serviceUnderTest);

Generates this SQL:

Microsoft.EntityFrameworkCore.Database.Command:Information: Executed DbCommand (646ms) [Parameters=[@__workingCompare_1='?' (Size = 4000)], CommandType='Text', CommandTimeout='30']
SELECT [ev].[Id], [ev].[Data], [ev].[EventFired], [ev].[EventTypeId], [ev].[IncidentId], [ev].[Occurred]
FROM [Events] AS [ev]
WHERE [ev].[EventTypeId] IN (1, 10) AND ((CHARINDEX(@__workingCompare_1, [ev].[Data]) > 0) OR (@__workingCompare_1 = N''))